### PR TITLE
Use a standard interface for all StyleColor kinds, and move them each to their own file

### PR DESCRIPTION
**Pull request by weinig**
#### d47a77a1cfa7e9f7cfee53030f702ebd659b5f54
<pre>
Use a standard interface for all StyleColor kinds, and move them each to their own file
<a href="https://bugs.webkit.org/show_bug.cgi?id=274005">https://bugs.webkit.org/show_bug.cgi?id=274005</a>

Reviewed by Matthieu Dubet and Tim Nguyen.

In preparation for adding support for relative colors to StyleColor, this
change adds explicit types for each StyleColor kind and gives each their own
files. This allows the implementation of StyleColor itself to be very simple,
just dispatching to whatever kind is resident in the variant.

The one complication here is that StyleColorMix (and any relative color
support type in the future), wants to have a StyleColor as a member, creating
a recursive relationship. This means we have be very careful not to instantiate
the ColorKind variant in the StyleColor header at all, as it would lead to
undefined type compile errors. Instead, we out-of-line all the operations.

With a bunch of StyleColor related files now existing, they also all got
moved to the color subdirectory.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Added new files, updated for file moves.

* Source/WebCore/css/color/CSSColorDescriptors.h:
    - Add missing #include.

* Source/WebCore/css/color/CSSColorMixSerialization.h:
    - Updated #include to use new StyleColorMix header.

* Source/WebCore/css/color/StyleAbsoluteColor.cpp: Added.
* Source/WebCore/css/color/StyleAbsoluteColor.h: Added.
    - Added. Simple wrapper around a Color.

* Source/WebCore/css/color/StyleColor.cpp: Renamed from Source/WebCore/css/StyleColor.cpp.
* Source/WebCore/css/color/StyleColor.h: Renamed from Source/WebCore/css/StyleColor.h.
    - Moved. Updated to out-of-line all uses of ColorKind
      and use generic dispatch in all possible places.

* Source/WebCore/css/color/StyleColorMix.cpp: Added.
* Source/WebCore/css/color/StyleColorMix.h: Added.
    - Moved from StyleColor.h/cpp

* Source/WebCore/css/color/StyleCurrentColor.cpp: Added.
* Source/WebCore/css/color/StyleCurrentColor.h: Added.
    - Moved from StyleColor.h/cpp

Canonical link: <a href="https://commits.webkit.org/278620@main">https://commits.webkit.org/278620@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/4e16d944c6dcbb9980048501190ab413fff73099

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51139 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30441 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3473 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54396 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1829 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53442 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36735 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1503 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/41647 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53238 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/28072 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44075 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/22763 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/25398 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1331 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47378 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1403 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55992 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26249 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/1298 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49047 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27497 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44142 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48184 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11183 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28378 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27228 "Built successfully") | | |

### CODE REVIEW

**Review by nt1m — APPROVED** [review]

File: Source/WebCore/css/color/CSSColorDescriptors.h [hunk 1]
```diff
@@ -26,8 +26,10 @@
 #pragma once
 
 #include "CSSPropertyParserConsumer+RawTypes.h"
+#include "CSSValueKeywords.h"
 #include "Color.h"
 #include "ColorTypes.h"
+#include <variant>
```
Comment (mdubet) [inline Source/WebCore/css/color/CSSColorDescriptors.h:32]: Do we need this include ?

File: Source/WebCore/css/color/StyleColor.cpp [hunk 2]
```diff
@@ -32,36 +32,105 @@
...
+StyleColor::StyleColor(StyleColor&&) = default;
+StyleColor& StyleColor::operator=(StyleColor&&) = default;
+StyleColor::~StyleColor() = default;
+
+bool StyleColor::operator==(const StyleColor&) const = default;
+
+// This helper allows us to treat all the alternatives in ColorKind
+// as const references, pretending the UniqueRefs don't exist.
+template<typename... F>
+decltype(auto) StyleColor::visit(const StyleColor::ColorKind& color, F&&... f)
```
Comment (mdubet) [inline Source/WebCore/css/color/StyleColor.cpp:98]: 👍

File: Source/WebCore/css/color/StyleCurrentColor.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/Forward.h>
```
Comment (mdubet) [inline Source/WebCore/css/color/StyleCurrentColor.h:28]: Do we need this include here?

**Review by mdubet — APPROVED** [review]

File: Source/WebCore/css/color/StyleCurrentColor.h (hunk 3 above)
Comment (weinig) [reply Source/WebCore/css/color/StyleCurrentColor.h:28]: Yep, for TextStream

File: Source/WebCore/css/color/CSSColorDescriptors.h (hunk 1 above)
Comment (weinig) [reply Source/WebCore/css/color/CSSColorDescriptors.h:32]: yep.

**Pull request merged**